enhancements/subscription-content/subscription-injection.md

```diff
@@ -0,0 +1,397 @@
+2. Create a share for the `entitlement` and `yumRepository` resources as declared in the Projected
+   Resource CSI Driver proposal.
+3. Create a `ClusterRole` associated with the bundle, which grants `get`, `watch`, and `list`
+   permissions to the associated shares.
+4. Aggregate the cluster role to a configurable list of system roles. By default this will be
+   aggregated to the `edit` and `admin` roles.
+5. Add owner references to the generated `ClusterRole` and shares. This ensures the dependent
+   objects are deleted when the `ClusterBundle` is deleted.
+
+#### Subscription Injection
+
+Developers can inject subscription bundles by adding the `subscription.openshift.io/cluster-bundle`
+annotation to a pod template spec. Any workload controller which ultimately creates a `Pod` will be
+supported. For example, to inject a subscription into a `Deployment`, use the following YAML:
+
+```yaml
+kind: Deployment
+apiVersion: v1
+metadata:
+  name: my-rhel-deployment
+spec:
+  replicas: 2
+  selector:
+    matchLabels:
+      app: "rhel-deployment"
+  template:
+    metadata:
+      annotations:
+        "subscription.openshift.io/cluster-bundle": rhel8
+      labels:
+        app: "rhel-deployment"
+    spec:
+      containers:
+      - name: run
+        image: "registry.redhat.io/ubi8/ubi:latest"
+        ...
+```
+
+The Subscription Injection mutating admission webhook will watch all newly created pods. The webhook
+will do the following:
+
+1. Look for the `subscription.openshift.io/cluster-bundle` annotation
+2. Read the `ClusterBundle` with a name matching the annotation value
+3. Add a `Volume` in the pod for the `entitlement` referenced in the bundle which sources data via
+   the projected resource CSI driver.
+4. Add a `Volume` in the pod for the `yumRepository` referenced in the bundle wihc sources data via
+   the projected resource CSI driver
+5. Mount the `entitlement` volume into `/run/secrets/etc-pki-entitlement` for each container in the
+   pod, including init and ephemeral containers.
+6. Mount the `yumRepository` volume into `/run/secrets` for each container in the pod, including
+   init and ephemeral containers.
+
+Per upstream docs, the mutating admission webhook will need to register itself with the cluster and
+run as a service in its own namespace [1]. The operator installing the Subscription Injection
+webhook will need to do the following:
+
+1. Create a `Deployment` that runs the webhook in a highly available mode.
+2. Create a `Service` to expose the webhook pods
+3. Create a `ConfigMap` where the service serving CA can be injected
+4. Create a `MutatingWebhookConfiguration` that fires the webhook for all created pods.
+5. When the service serving CA in item 3 is injected or updated, the operator needs to update the
+   caBundle provided to the webook configuration in 4.
+
+[1]
+https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/#mutating-webhook-configuration
```

Comment:
Use of https://github.com/openshift/generic-admission-server should be considered ... either as the primary choice, or listed in the alternatives section as something that was dismissed

Reply:
Interesting - will add as an implementation path.

Reply:
@gabemontero so I took a look at that - it's a deployment option, but I don't see it being adopted by other OpenShift admission webhooks.

Examples of active mutating admission webhooks:

* https://github.com/openshift/cluster-resource-override-admission

Reply:
yeah @deads2k would provide a better sales job than I could (assuming it should  still be sold) .... I don't recall if https://github.com/openshift/kubernetes-namespace-reservation is used directly or if it served as the basis for equivalent function elsewhere